// slnodejs/tsOutputs/common/http/backend-proxy-with-default-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BackendProxyFactory = void 0;
const backend_proxy_1 = require("./backend-proxy");
const http_client_1 = require("./http-client");
const metrics_store_factory_1 = require("../metrics-process/metrics-store-factory");
class BackendProxyFactory {
    static initWithDefaultClient(agentInstanceData, config, logger) {
        const metadata = {
            labId: config.labId,
            appName: config.appName,
            branchName: config.branchName,
            buildName: config.buildName,
            agentId: agentInstanceData.agentId,
            buildSessionId: config.buildSessionId,
            agentType: agentInstanceData.agentType,
            agentTechnology: agentInstanceData.technology,
            agentVersion: agentInstanceData.agentVersion,
            testProjectId: config.testProjectId,
            targetTestProjectId: config.targetTestProjectId,
        };
        const httpMetricsStore = metrics_store_factory_1.HttpMetricsStoreFactory.create(logger);
        const client = new http_client_1.HttpClient(config, metadata, logger, httpMetricsStore);
        return new backend_proxy_1.BackendProxy(agentInstanceData, config, logger, client);
    }
}
exports.BackendProxyFactory = BackendProxyFactory;
