// slnodejs/tsOutputs/cli-parse/executors/instrument-args-dto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstrumentationArgsDTO = void 0;
/**
 * Instrumenter takes the given arg object and parse it to json.
 * The arg received is a 'Commander' object which cannot be parsed due to circular error.
 * This is a class that implements the IBuildInstrumentationArgs with all members as undefined in order to
 * get all the keys by 'Object.keys' in {@toDto}
 */
class InstrumentationArgsDTO {
    constructor() {
        this.buildsessionid = undefined;
        this.buildsessionidfile = undefined;
        this.es6Modules = undefined;
        this.labid = undefined;
        this.logsUrl = undefined;
        this.outputPath = undefined;
        this.skipUpgrade = undefined;
        this.proxy = undefined;
        this.token = undefined;
        this.tokenfile = undefined;
        this.workspacePath = undefined;
        this.collectorUrl = undefined;
        this.enableOpenTelemetry = false;
        this.allowCORS = undefined;
        this.footprintsMapping = undefined;
        this.overrideServer = undefined;
        this.excludeTokenFromBrowserConfig = undefined;
        this.useRelativeSlMapping = undefined;
        this.newInstrumenter = undefined;
        this.resolveWithoutHash = undefined;
        this.delayShutdownInSeconds = undefined;
        this.removeSensitiveData = undefined;
        this.reduceFileSize = undefined;
        this.ctxPropagationOnly = false;
        this.splitPreambleIntoFile = false;
        this.preambleFileName = undefined;
        this.preambleFileUrl = undefined;
        this.toggleBasedBrowserInstrumentation = false;
    }
    static create(args) {
        const result = {};
        const iBuildArgs = new InstrumentationArgsDTO();
        const keys = Object.keys(iBuildArgs);
        keys.forEach((key) => { var _a; return (_a = result[key]) !== null && _a !== void 0 ? _a : (result[key] = args[key]); });
        return result;
    }
}
exports.InstrumentationArgsDTO = InstrumentationArgsDTO;
